migrate to settings plugins and eliminate variant-based conditional compilation

bcressey opened the pull request:
**Issue number:**
Closes: #3615 

**Description of changes:**

Some major long-term efforts culminate here:
1. Move all the distinct settings models into settings plugins.
2. Make apiserver / datastore / models work with settings plugins.
3. Move all the distinct variant defaults into settings defaults crates.
4. Teach storewolf to load defaults at runtime.
5.  Add new packages and dependencies to pull in the correct plugin and defaults for each variant.

The end result of this is that `os` is no longer compiled differently based on the variant being built, and everything still works.

**Testing done:**
Built `aws-dev` and `aws-k8s-1.28-nvidia`. Verified that they used the expected defaults and supported the correct settings.


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

sam-berning committed:
add settings-plugins package

Signed-off-by: Sam Berning <bernings@amazon.com>

sam-berning committed:
datastore: allow u64 serialization

Settings are first serialized to JSON values before being transformed
into the final type by the settings plugin. Number fields in the JSON
value show up as "u64" by default. Allow them to pass through and let
serialization by the settings plugin deal with numbers outside the
range of what the actual type expects.

Signed-off-by: Sam Berning <bernings@amazon.com>

sam-berning committed:
apiserver: add "settings" prefix

When interacting with the wrapper types provided by settings plugins,
the actual struct name is not available to serde. When serializing,
add the "settings" prefix, and when deserializing, treat an unnamed
struct as "settings" instead of an error.

Signed-off-by: Sam Berning <bernings@amazon.com>

Commit:
packages: remove packaging for settings extensions

These aren't currently built, and there are no immediate plans to use
the resulting binaries.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
os: depend on settings plugins

Various programs like `apiserver` now require a settings plugin in
order to function instead of immediately panicking. Add a requirement
to ensure that some plugin is installed. Which one doesn't matter as
long as it implements the plugin interface.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-plugins: add aws-ecs-1 plugin

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-plugins: add aws-k8s plugin

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-plugins: add metal-k8s plugin

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-plugins: add vmware-dev plugin

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-plugins: add vmware-k8s plugin

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
apiserver: fix test cases for settings plugins

Add a simple settings plugin to expose the fields that the existing
unit tests expect, and an accessor macro to peel apart the serialized
form of the otherwise-opaque type.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
models: remove structs that migrated to extensions

Signed-off-by: Ben Cressey <bcressey@amazon.com>

sam-berning committed:
models: implement Default and Clone for types

Types used by a settings plugin must implement Default and Clone.

Signed-off-by: Sam Berning <bernings@amazon.com>

Commit:
storewolf: read defaults from system at runtime

Instead of merging TOML snippets at build time and saving the result
in the binary, read the desired TOML from the filesystem.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
os: depend on settings defaults

Now that storewolf no longer has a built-in copy of the merged TOML
file to initialize the datastore, it must load it at runtime instead.
Add a requirement to ensure that a package providing those defaults
is installed.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
add settings-defaults package

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
models: relocate shared defaults

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-defaults: add defaults for aws-dev

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-defaults: add defaults for aws-ecs-1

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-defaults: add defaults for aws-ecs-1-nvidia

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-defaults: add defaults for aws-ecs-2-nvidia

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-defaults: add defaults for aws-ecs-2

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-defaults: add defaults for aws-k8s-1.24-nvidia

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-defaults: add defaults for aws-k8s-1.24

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-defaults: add defaults for aws-k8s-1.25-nvidia

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-defaults: add defaults for aws-k8s-1.25

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-defaults: add defaults for aws-k8s-1.26-nvidia

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-defaults: add defaults for aws-k8s-1.30

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-defaults: add defaults for metal-dev

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-defaults: add defaults for metal-k8s-1.30

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-defaults: add defaults for aws-k8s-1.26

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-defaults: add defaults for vmware-dev

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-defaults: add defaults for aws-k8s-1.30-nvidia

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-defaults: add defaults for vmware-k8s-1.30

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
models: drop build-time dependency on VARIANT

With settings moved to plugins and defaults moved to a TOML file, the
variant we're building for no longer affects the compiled code.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
os: drop variant sensitive flag

The software packaged in `os` is no longer affected by the variant,
so turn it into a regular package.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
sources: refresh settings sdk dependencies

Signed-off-by: Ben Cressey <bcressey@amazon.com>

sam-berning committed:
models: load settings through plugin

Instead of compiling "models" with a different, per-variant settings
structure, load a settings plugin at runtime. This removes one of the
last major sources of conditional compilation for first-party Rust
programs.

Signed-off-by: Sam Berning <bernings@amazon.com>

cbgbt commented on sources/api/apiserver/src/server/controller.rs:
```diff
@@ -366,8 +366,8 @@ pub(crate) fn set_settings<D: DataStore>(
     trace!("Serializing Settings to write to data store");
-    let pairs =
-        to_pairs(settings).context(error::DataStoreSerializationSnafu { given: "Settings" })?;
+    let pairs = to_pairs_with_prefix("settings", settings)
+        .context(error::DataStoreSerializationSnafu { given: "Settings" })?;
```
Just making sure I understand -- this is required because the models were previously called `Settings` and the serializer would subsequently add that as part of each K/V name?

bcressey replied:
Yes - now the name of the struct is not visible to the serializer, so we have to provide it. I feel like there's probably a way to make the old behavior work, but I couldn't find it in the time I spent trying.

cbgbt commented on sources/api/datastore/src/serialization/pairs.rs:
```diff
@@ -214,11 +218,6 @@ impl<'a> ser::Serializer for Serializer<'a> {
     }
 
     // Types we can't (or don't want to) represent.
-    // Can't fit u64 into signed 64-bit range.
-    fn serialize_u64(self, _v: u64) -> Result<()> {
-        bad_type("u64")
-    }
-
```
This is a bit concerning, but I probably don't understand the source of this constraint.

My assumption is that `u64` is disallowed due to a constraint in `serde_json` (indeed, Javascript cannot store integers larger than 53bits). Looking at [`serde_json` docs](https://docs.rs/serde_json/latest/serde_json/value/struct.Number.html#method.as_u64), it seems like maybe it does support `u64` though. Are we making some change to the datastore which makes this more flexible?

bcressey replied:
> Are we making some change to the datastore which makes this more flexible?

On the datastore side, the "no u64" constraint seems to have come out of [this discussion](https://github.com/bottlerocket-os/bottlerocket/issues/260#issuecomment-543314276). Not sure if `serde_json` didn't support `u64` at the time, or if Tom was concerned about not tying persisted data to behavior outside the actual JSON spec.

It's still the case that we don't want to actually persist or support u64s in the datastore - if nothing else because the TOML serde implementation doesn't support them.

The problem being fixed here is that we need to bounce conversions through a serde_json Value, so that our custom serializer gets structured data to work with rather than the serde_json RawValue that comes out of the plugins. And when the serializer sees a `Value::Number(n)` it takes the `serialize_u64` code path.

I've confirmed that the API still rejects larger numbers, so we're not actually losing the "no u64" property:
```
# apiclient set updates.seed=5000000000
Failed to change settings: Failed PATCH request to '/settings/keypair?tx=apiclient-set-IzntMxvwbPebZ0GV': Status 400 when PATCHing /settings/keypair?tx=apiclient-set-IzntMxvwbPebZ0GV: Unable to match your input to the data model.  We may not have enough type information.  Please try the --json input form.  Cause: Error during deserialization: invalid value: integer `5000000000`, expected u32 at line 1 column 29

# apiclient apply <<EOF
> [settings.updates]
> seed = 5000000000
> EOF
Failed to apply settings: Failed to PATCH settings from '-' to '/settings?tx=apiclient-apply-WEVNJ04u1F1d51xU': Status 400 when PATCHing /settings?tx=apiclient-apply-WEVNJ04u1F1d51xU: Json deserialize error: invalid value: integer `5000000000`, expected u32 at line 1 column 29
```

cbgbt commented on packages/settings-plugins/settings-plugins.spec:
```diff
@@ -0,0 +1,205 @@
+BuildRequires: %{_cross_os}glibc-devel
+Requires: %{_cross_os}glibc-devel
+
+Requires: %{_cross_os}settings-plugin(any)
```
I assume we'll need hoist this or something like it into `os` to force kits to provide something?

I suppose alternatively (and maybe correctly) we provide a null plugin.

bcressey replied:
> I assume we'll need hoist this or something like it into os to force kits to provide something?

I added the dependency to `os.spec` in a later. However, your comment made me realize that in the absence of a "preferred" choice defined by requires in the top-level package, the approach here would allow `dnf` to pick a settings plugin at random to satisfy the `%{_cross_os}settings-plugin(any)` dependency. (And same for defaults.)

I'll tighten that up in the next revision.

cbgbt commented on packages/settings-defaults/settings-defaults.spec:
```diff
@@ -0,0 +1,360 @@
+%{summary}.
+
+%package aws-k8s-1.24
+Summary: Settings defaults for the aws-k8s-1.24 variant
```
Worth mentioning `1.23` in the summary? (Same Q for the nvidia defaults for these versions)

cbgbt reviewed (commented):
Clever use of dynamic libraries to solve this problem. Nice!

webern reviewed (approved):
I'll keep looking through this but don't want to be a blocker 😅

This is exciting!

webern commented on sources/settings-plugins/dummy/Cargo.toml:
```diff
@@ -0,0 +1,34 @@
+[package]
+name = "settings-plugin-dummy"
```
This could use a readme or something to explain its existence.

webern commented on sources/settings-defaults/metal-k8s-1.30/build.rs:
```diff
@@ -0,0 +1,3 @@
+fn main() {
+    bottlerocket_defaults_helper::generate_defaults_toml().unwrap();
```
It seems like this could be deduplicated by placing it at `../build.rs`.

webern commented on sources/settings-defaults/metal-dev/src/lib.rs:
```diff
@@ -0,0 +1 @@
+// This "crate" only generates a defaults.toml as a side effect of the build.
```
It seems like this could be deduplicated by placing it at `../lib.rs`.

webern commented on sources/models/src/variant/.keep:
```diff
@@ -1,3 +1,2 @@
-We want git to track this directory so that it exists for end-user builds.
-It's needed by build.rs and the symlinks described in README.md.
-It's clearer what's going on if the directory exists already, rather than build.rs creating it.
+FIXME: this directory is no longer needed for builds, but can't be removed yet
+because Twoliter expects it to exist.
```
Yeah, it's `build.Dockerfile` that needs it.

webern commented on sources/models/build.rs:
```diff
@@ -1,130 +1,3 @@
-use std::process;
-
-/// We create a link from 'current' to the variant selected by the environment variable above.
-const VARIANT_LINK: &str = "src/variant/current";
```
:tada:

Commit:
settings-defaults: add shared Rust code

Each settings-defaults crate will use the same build script and the same
empty library file.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

bcressey commented:
I still need to rebase, but the above force push uses the tagged `v0.1.0` release of `bottlerocket-settings-sdk`, and addresses feedback so far:
* drop `authors` field from the new crate manifests (@sam-berning)
* use shared sources for the `shared-defaults` crates (@webern )
* rename `settings-plugin-dummy` to `simple-settings-plugin`, move under `api`, add README (@webern)
* revamp dependencies in the `settings-defaults` package to force `dnf` to pick an exact match (@cbgbt)
* drop `settings-extensions` packages since they aren't used and don't have a clear transition path to out-of-tree

Commit:
settings-plugins: add aws-dev plugin

Signed-off-by: Sam Berning <bernings@amazon.com>
Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-plugins: add aws-ecs-2 plugin

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
settings-plugins: add metal-dev plugin

Signed-off-by: Ben Cressey <bcressey@amazon.com>
